renderdoc/driver/d3d12/d3d12_shaderdebug.cpp

```diff
@@ -96,8 +96,176 @@ bool D3D12DebugAPIWrapper::CalculateMathIntrinsic(DXBCBytecode::OpcodeType opcod
+  uavParam.ParameterType = D3D12_ROOT_PARAMETER_TYPE_UAV;
+  uavParam.ShaderVisibility = D3D12_SHADER_VISIBILITY_ALL;
+  uavParam.Descriptor.ShaderRegister = 0;
+  uavParam.Descriptor.RegisterSpace = 0;
+  uavParam.Descriptor.Flags = D3D12_ROOT_DESCRIPTOR_FLAG_NONE;
+  rootSig.Parameters.push_back(uavParam);
+
+  ID3DBlob *root = m_pDevice->GetShaderCache()->MakeRootSig(rootSig);
+  if(root == NULL)
+  {
+    RDCERR("Failed to create root signature for pixel shader debugging");
+    SAFE_RELEASE(csBlob);
+  }
+  ID3D12RootSignature *pRootSignature = NULL;
+  HRESULT hr =
+      m_pDevice->CreateRootSignature(0, root->GetBufferPointer(), root->GetBufferSize(),
+                                     __uuidof(ID3D12RootSignature), (void **)&pRootSignature);
+  SAFE_RELEASE(root);
+  if(FAILED(hr))
+  {
+    RDCERR("Failed to create root signature for pixel shader debugging HRESULT: %s",
+           ToStr(hr).c_str());
+    SAFE_RELEASE(csBlob);
+    return false;
+  }
+
+  // Create a fresh PSO, since the shader being debugged could be using the graphics pipe
+  D3D12_COMPUTE_PIPELINE_STATE_DESC psoDesc;
+  psoDesc.pRootSignature = pRootSignature;
+  psoDesc.CS.BytecodeLength = csBlob->GetBufferSize();
+  psoDesc.CS.pShaderBytecode = csBlob->GetBufferPointer();
+  psoDesc.NodeMask = 0;
+  psoDesc.CachedPSO.pCachedBlob = nullptr;
+  psoDesc.CachedPSO.CachedBlobSizeInBytes = 0;
+  psoDesc.Flags = D3D12_PIPELINE_STATE_FLAG_NONE;
+
+  ID3D12PipelineState *pPipeState = NULL;
+  hr = m_pDevice->CreateComputePipelineState(&psoDesc, __uuidof(ID3D12PipelineState),
+                                             (void **)&pPipeState);
+  SAFE_RELEASE(csBlob);
+  if(FAILED(hr))
+  {
+    RDCERR("Failed to create PSO for shader debugging HRESULT: %s", ToStr(hr).c_str());
+    SAFE_RELEASE(pRootSignature);
+    return false;
+  }
+
+  // Create buffer to store computed result
+  D3D12_RESOURCE_DESC rdesc;
+  ZeroMemory(&rdesc, sizeof(D3D12_RESOURCE_DESC));
+  rdesc.Dimension = D3D12_RESOURCE_DIMENSION_BUFFER;
+  rdesc.Width = sizeof(float) * 8;    // Output buffer is 2x float4
+  rdesc.Height = 1;
+  rdesc.DepthOrArraySize = 1;
+  rdesc.MipLevels = 1;
+  rdesc.Format = DXGI_FORMAT_UNKNOWN;
+  rdesc.Flags = D3D12_RESOURCE_FLAG_ALLOW_UNORDERED_ACCESS;
+  rdesc.Layout = D3D12_TEXTURE_LAYOUT_ROW_MAJOR;
+  rdesc.SampleDesc.Count = 1;
+  rdesc.SampleDesc.Quality = 0;
+
+  D3D12_HEAP_PROPERTIES heapProps;
+  heapProps.Type = D3D12_HEAP_TYPE_DEFAULT;
+  heapProps.CPUPageProperty = D3D12_CPU_PAGE_PROPERTY_UNKNOWN;
+  heapProps.MemoryPoolPreference = D3D12_MEMORY_POOL_UNKNOWN;
+  heapProps.CreationNodeMask = 1;
+  heapProps.VisibleNodeMask = 1;
+
+  ID3D12Resource *pResultBuffer = NULL;
+  hr = m_pDevice->CreateCommittedResource(&heapProps, D3D12_HEAP_FLAG_NONE, &rdesc,
+                                          D3D12_RESOURCE_STATE_UNORDERED_ACCESS, NULL,
+                                          __uuidof(ID3D12Resource), (void **)&pResultBuffer);
```

Comment:
Similarly this result buffer could be created once and kept around. I think resource creation is likely more expensive in D3D12 since the runtime/driver probably won't do caching and suballocation for us, so since this function is _reasonably_ high frequency it would be better to avoid creations where we can easily help it.

Reply:
Hmm in that case we could store the PSOs for each opcode handled here and avoid nearly all per-instruction resource creation. I'll look into that.

Reply:
You could also have an extra input parameter to the shader with the opcode and have the code do if/else for each one. That's something I should have done for D3D11 to be honest.